rust/src/accumulator.rs

```diff
@@ -0,0 +1,224 @@
+/// more can be added.
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum UsageUnit {
+    Milliseconds,
+    Bytes,
+    MillisecondsSec,
+}
+
+impl fmt::Display for UsageUnit {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            UsageUnit::Milliseconds => write!(f, "milliseconds"),
+            UsageUnit::Bytes => write!(f, "bytes"),
+            UsageUnit::MillisecondsSec => write!(f, "milliseconds_sec"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct UsageKey {
+    pub quantized_timestamp: DateTime<Local>,
+    pub resource_id: String,
+    pub app_feature: String,
+    pub unit: UsageUnit,
+}
+
+pub struct UsageAccumulator {
+    usage_batch: HashMap<UsageKey, u64>,
+    granularity_sec: Duration,
+    first_timestamp: Option<DateTime<Local>>,
+}
+
+impl UsageAccumulator {
+    /// Constructs a new Accumulator. Here is where the granularity
+    /// is provided.
+    pub fn new(granularity_sec: Option<Duration>) -> Self {
+        Self {
+            usage_batch: HashMap::new(),
+            granularity_sec: granularity_sec.unwrap_or(Duration::seconds(60)),
+            first_timestamp: None,
+        }
+    }
+
+    /// Records an amount of usage for a resource, app_feature, timestamp
+    /// tuple.
+    ///
+    /// The timestamp provided is then quantized according to the
+    /// granularity this structure is instantiated with so data is
+    /// bucketed with fixed bucket sizes.
+    /// The system timestamp should be passed in most cases.
+    pub fn record(
+        &mut self,
+        usage_time: DateTime<Local>,
+        resource_id: &str,
+        app_feature: &str,
+        amount: u64,
+        usage_unit: UsageUnit,
+    ) {
+        let quantized_timestamp: DateTime<Local> =
+            usage_time.duration_trunc(self.granularity_sec).unwrap();
+
+        if self.first_timestamp.is_none() {
+            self.first_timestamp = Some(quantized_timestamp);
+        }
+
+        let key = UsageKey {
+            quantized_timestamp,
+            resource_id: resource_id.to_string(),
+            app_feature: app_feature.to_string(),
+            unit: usage_unit,
+        };
+
+        let value = self.usage_batch.entry(key).or_default();
+        *value += amount;
+    }
+
+    /// Returns true if the bucket is ready to be flushed.
+    ///
+    /// Ready to be flushed means that the bucket is not empty
+    /// and at least `granularity_sec` seconds have passed since
+    /// the first chunk of data was added.
+    pub fn should_flush(&self, current_time: DateTime<Local>) -> bool {
+        return self.first_timestamp.is_some()
+            && self.usage_batch.keys().len() > 0
+            && current_time - self.first_timestamp.unwrap() > self.granularity_sec;
+    }
+
+    /// Return the current bucket and clears up the state.
+    pub fn flush(&mut self) -> HashMap<UsageKey, u64> {
```

Comment:
Should this be private? Otherwise it's unclear whether you expect the user to call this function or not

Reply:
The user does not interact with the accumulator. 
They interact only with the Accountant. The Accountant uses the Accumulator. 
Actually I think I should make the entire accumulator module not public.